In [ ]:
pip install neat-python

In [ ]:
import neat

# Define a fitness function
def eval_genomes(genomes, config):
    for genome_id, genome in genomes:
        net = neat.nn.FeedForwardNetwork.create(genome, config)
        fitness = 0

        # Example: Evaluate the network on some test cases
        test_cases = [
            ([0, 0], 0),
            ([0, 1], 1),
            ([1, 0], 1),
            ([1, 1], 0)
        ]

        for inputs, expected_output in test_cases:
            output = net.activate(inputs)[0]
            fitness -= (output - expected_output) ** 2

        genome.fitness = fitness

# Load the configuration file
config_path = '/content/neat-config'
config = neat.config.Config(neat.DefaultGenome, neat.DefaultReproduction,
                            neat.DefaultSpeciesSet, neat.DefaultStagnation,
                            config_path)

# Create the population
population = neat.Population(config)

# Add reporters to show progress
population.add_reporter(neat.StdOutReporter(True))
stats = neat.StatisticsReporter()
population.add_reporter(stats)

# Run NEAT
winner = population.run(eval_genomes, n=50)

print(f'Best genome:\n{winner}')



 ****** Running generation 0 ****** 

Population's average fitness: -2.00000 stdev: 0.00000
Best fitness: -2.00000 - size: (1, 0) - species 1 - id 1
Average adjusted fitness: 0.000
Mean genetic distance 0.541, standard deviation 0.456
Population of 150 members in 1 species:
   ID   age  size  fitness  adj fit  stag
  ====  ===  ====  =======  =======  ====
     1    0   150     -2.0    0.000     0
Total extinctions: 0
Generation time: 0.010 sec

 ****** Running generation 1 ****** 

Population's average fitness: -1.96949 stdev: 0.11723
Best fitness: -1.08316 - size: (1, 1) - species 1 - id 196
Average adjusted fitness: 0.031
Mean genetic distance 1.111, standard deviation 0.609
Population of 150 members in 1 species:
   ID   age  size  fitness  adj fit  stag
  ====  ===  ====  =======  =======  ====
     1    1   150     -1.1    0.031     0
Total extinctions: 0
Generation time: 0.020 sec (0.015 average)

 ****** Running generation 2 ****** 

Population's average fitness: -1.88833 stde

Congfig File:

[NEAT]
pop_size = 150
fitness_criterion = max
fitness_threshold = 3.5
no_fitness_termination = True
reset_on_extinction = True

[DefaultGenome]
num_inputs = 2
num_outputs = 1
num_hidden = 0
feed_forward = True
activation_default = sigmoid
activation_mutate_rate = 0.0
aggregation_default = sum
aggregation_mutate_rate = 0.0
bias_init_mean = 0.0
bias_init_stdev = 1.0
bias_max_value = 30.0
bias_min_value = -30.0
bias_mutate_power = 0.5
bias_mutate_rate = 0.7
bias_replace_rate = 0.1
response_init_mean = 1.0
response_init_stdev = 0.0
response_max_value = 30.0
response_min_value = -30.0
response_mutate_power = 0.0
response_mutate_rate = 0.0
response_replace_rate = 0.0
weight_init_mean = 0.0     
weight_init_stdev = 1.0     
weight_max_value = 30.0    
weight_min_value = -30.0   
weight_mutate_power = 0.5     
weight_mutate_rate = 0.8     
weight_replace_rate = 0.1   
enabled_default = True
enabled_mutate_rate = 0.01
conn_add_prob = 0.5
conn_delete_prob = 0.5
node_add_prob = 0.2
node_delete_prob = 0.2

# Genome compatibility parameters
compatibility_disjoint_coefficient = 1.0
compatibility_weight_coefficient = 0.5

[DefaultSpeciesSet]
compatibility_threshold = 3.0

[DefaultStagnation]
species_fitness_func = max
max_stagnation = 20
species_elitism = 2

[DefaultReproduction]
elitism = 2
survival_threshold = 0.2